tls: support BoringSSL private key async functionality

ipuustin opened the pull request:
This PR adds BoringSSL private key API abstraction, as discussed in https://github.com/envoyproxy/envoy/issues/6248. All comments and discussion is welcomed to get the API sufficient for most private key API tasks.

The PR contains the proposed API and the way how it can be used from `ssl_socket.h`. Also there is some code showing how the `PrivateKeyMethodProvider` is coming from TLS certificate config. Two example private key method providers are included in the tests.

Description: tls: support BoringSSL private key async functionality
Risk Level: medium
Testing: two basic private key provider implementation
Docs Changes: TLS arch doc, cert.proto doc

rojkov commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -60,6 +60,15 @@ void SslSocket::setTransportSocketCallbacks(Network::TransportSocketCallbacks& c
+    ops_ = provider_->getPrivateKeyOperations(*this, callbacks_->connection().dispatcher());
+    Ssl::PrivateKeyMethodSharedPtr private_key_methods = ops_->getPrivateKeyMethods(ssl_.get());
+    if (private_key_methods) {
+      SSL_set_private_key_method(ssl_.get(), private_key_methods.get());
```
I guess this function should be incapsulated somewhere inside `source/extensions/transport_socket_tls/private_key` to avoid messing with OpenSSL's API if we add conditional OpenSSL builds.

ipuustin replied:
You're right -- the setup could also be in a `PrivateKeyImpl` class or similar. On the other hand the whole concept of BoringSSL private key calls doesn't make sense if OpenSSL is used.

ipuustin commented:
@htuch @PiotrSikora I added a second patch which adds support for registering private key provider factories. Any comments whether this is going the right way would be helpful. I was considering between this approach and extending `Connection` API to have network read filters to be able to access SSL internal structures; WDYT?

htuch commented on api/envoy/api/v2/auth/cert.proto:
```diff
@@ -147,6 +147,11 @@ message TlsSessionTicketKeys {
 
+message PrivateKeyOperations {
+  string private_key_provider = 1;
+  core.ConfigSource private_key_provider_config = 2;
```
Starting from first principles on this review (just noticed it when you tagged me). How come we need to use `ConfigSource` here? Wouldn't the correct thing be a typed config, i.e. an `Any`, similar to how we configure other extensions? I think the best example for what you're trying to integrate here is something like the gRPC credential extension, see https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/google_grpc_creds_impl.h

ipuustin replied:
That makes sense. I added a patch which does just that. I was bit unsure if the plan is to use the combined Struct/Any approach for new extension types, or just go with Any.

I'm currently thinking that the configuration conversion would be done in the private key provider extensions something like this:

```c++
Ssl::PrivateKeyOperationsProviderSharedPtr TestPrivateKeyOperationsFactory::createPrivateKeyOperationsProviderInstance(
    const envoy::api::v2::auth::PrivateKeyOperations& message,
    Server::Configuration::TransportSocketFactoryContext& private_key_provider_context) {
  ProtobufTypes::MessagePtr proto_config = std::make_unique<test::TestPrivateKeyOperationsProto>();

  Config::Utility::translateOpaqueConfig(message.typed_config(), ProtobufWkt::Struct(), *proto_config);
  const test::TestPrivateKeyOperationsProto config = MessageUtil::downcastAndValidate<const test::TestPrivateKeyOperationsProto&>(*proto_config);

  return std::make_shared<TestPrivateKeyOperationsProvider>(config);
}
```

htuch replied:
Yeah, you can factor some of this out into an extension factory base, you can see what was done for L7 filters at https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/common/factory_base.h.

htuch commented:
Some answers to questions above:

> Validate the thread model. Now there is an unique PrivateKeyOperations instance for each connection, but the PrivateKeyOperationsProvider is shared. This makes the dispatcher model easier and the PrivateKeyOperations lifecycle can be tied with the caller.

Yep, this is how to do things in Envoy. It's fine to share the provider, assuming it's essentially a stateless (or immutable) factory, and all the private key operation state lives thread-local.

> How to get the private key to the provider? Just let the provider use SSL_get_privatekey()?

Isn't this highly dependent on the extension? Won't some have their private key material living in an HSM?

> Does SDS require any special handling?

SDS is orthogonal. It's inherently about fetching keys from a remote server and then using them for crypto, rather than delegating to dedicated crypto hardware. Maybe in theory you could imagine SDS being implemented behind this new interface, but I think what we have today works well in parallel to offload.

> Should we expose BoringSSL primitives (such as SSL *ssl) in the API?

Ideally what is being built works for both OpenSSL and BoringSSL, or some abstraction exists that can support this. I have no idea of the varying offload implementations, but just throwing this out there.

htuch commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -124,6 +133,21 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
 }
 
+void SslSocket::complete(Envoy::Ssl::PrivateKeyOperationStatus status) {
+  if (status == Envoy::Ssl::PrivateKeyOperationStatus::Success) {
```
What kind of threading guarantees does BoringSSL or OpenSSL provide? Do we know that the thread that private key operations are run on is the same as the invoking thread? Should `complete` post to the dispatcher to continue, or is that the job of the extension? This is one place that I'm not 100% sure of the thread correctness.

ipuustin replied:
The BoringSSL operations are driven by the caller: BoringSSL's private key complete function will be called when the `SSL_do_handshake()` is called the second time. However, the `complete()` function (indicating that the asynchronous operation is ready and `ssl_socket.cc` needs to call the handshake again) might in principle be called from whatever thread the private key operations extension happens to use. I believe the correct semantics is to define this to be an error, and expect the extension to call the `complete()` function from the worker thread context.

Should we add `ASSERT(isThreadSafe());` to `SslSocket::complete()` the way it is done in `dispatcher_impl.cc`?

I just wrote a (rough) implementation of a QAT private key extension to see if it fits the private key API. There I'm using a pipe and dispatcher to make sure that the callback thread is correct, because QAT device polling is done from a different thread. See the code here: https://github.com/ipuustin/envoy/blob/qat-provider/source/extensions/private_key_operations_providers/qat/qat_private_key_provider.cc#L20

htuch replied:
Yeah, I think it should be an error to `SslSocket::complete()` on a different thread. The Envoy pattern for handling this is to do a post to the original thread's dispatcher with the completion as the body. The only cross-thread safe dispatcher operation is `post()`.

ipuustin commented on include/envoy/ssl/context_config.h:
```diff
@@ -69,6 +70,12 @@ class ContextConfig {
   virtual void setSecretUpdateCallback(std::function<void()> callback) PURE;
+
+  /**
+   * @return the configured BoringSSL private key operations provider.
```
It doesn't - private key operations are a BoringSSL concept. OpenSSL uses a bit similar mechanism called asynchronous mode, which makes almost all operations potentially asynchronous. It's up to the backend (an OpenSSL engine) to implement the operations in an asynchronous way. OpenSSL provides a file descriptor to tell when the operation is ready.

The benefit of OpenSSL model is that many more operations can be performed asynchronously and also the possibility to have dynamic shared objects (engines) with a specified ABI to run the asynchronous operations. Also, if an asynchronous engine isn't used, the library falls back to synchronous operation. The drawback is that the OpenSSL library needs to be dynamically linked to Envoy in order to load external engines. Also, the `ASYNC` API is not present in BoringSSL.

I wrote a patch on top of @bdecoste's OpenSSL Envoy patches some time back to enable the OpenSSL asynchronous mode:
https://github.com/ipuustin/envoy/commit/41bc5bec35bb45416fa118f58f498e0ed04bf4e2

htuch replied:
Is there any hope of unifying these behind a common Envoy abstraction? That would be ideal, but obviously not if it needs to be a mega-project to make happen. Thoughts?

ipuustin replied:
It's my understanding that the BoringSSL and OpenSLL support will be implemented as separate TLS extensions. This means that the way the library is used will stay "hidden" in the implementations. I think the abstraction could be done in two different possible scopes:

1. Write support to loading OpenSSL Engines to OpenSSL TLS extension. Abstract the configuration and callback interface.
2. Find functions corresponding to the BoringSSL private key ops from OpenSSL ASYNC API, and provide a way to do "OpenSSL mini-engines" which would have the same functionality as the BoringSSL private key operations.

The problem with the first item is that it's already pretty well abstracted: there is a bunch of OpenSSL engines out there, and the benefit of Envoy knowing about them would be mainly to pass them configuration values. Envoy can use the dynamic engines quite easily (the only requirement is linking the OpenSSL library dynamically into Envoy), and there is `Dispatcher` access from the `Connection` interface to the TLS extension to facilitate the async event channel. The OpenSSL engines can also be configured from outside Envoy, by using OpenSSL configuration files (see https://www.openssl.org/docs/man1.1.1/man5/config.html). Writing the OpenSSL TLS extension to use asynchronous mode from day one would be a good idea.

The second item would mean in a sense that there would be feature parity between OpenSSL and BoringSSL implementations regarding private key operations. The OpenSSL would be pretty handicapped in this case, since the async engine support is more complete than BoringSSL private key operations. Using the QAT HW acceleration as an example, OpenSSL QAT engine (https://github.com/intel/QAT_Engine) can accelerate also EDCHE and some crypto operations, while the BoringSSL private key support is limited to RSA decryption and RSA+ECDSA signing. I think people would gravitate using the full OpenSSL engines rather than writing the Envoy-specific OpenSSL "mini-engines".

TL;DR: it feels to me that having a common abstraction for OpenSSL engines and BoringSSL private key operations might not have enough benefits compared to letting the TLS extensions set up the asynchronous operations as they please. However, I've had only limited exposure to OpenSSL engines, so it could be there are other considerations that I'm not aware of. Maybe there would be a special Envoy configuration Protobuf type for the TLS extensions, so that they could be configured separately? This could mean OpenSSL protobuf configuration message having a submessage for setting up the requested OpenSSL engine(s).

htuch replied:
@ipuustin thanks for the analysis. My thinking was that if X wanted to write a new private key offload extension Y for X's HSM, then it would be nice to have Y work no matter whether you go with BoringSSL or OpenSSL TLS. If there are already a number of OpenSSL engines out there that only require a light wrapping for Envoy, and OpenSSL is strictly more expressive than BoringSSL, I think there's not much hope or point of reuse.

So, I think this is resolved, we go with your current approach. Thanks again for taking the time to provide such a detailed discussion.

ipuustin commented on source/extensions/transport_sockets/tls/context_config_impl.cc:
```diff
@@ -104,6 +104,20 @@ getCertificateValidationContextConfigProvider(
+
+  const auto private_key_operations_config = config.private_key_operations();
+
+  if (private_key_operations_config.provider_name() != "") {
```
Ok, added a patch to fix this.

htuch commented on api/envoy/api/v2/auth/cert.proto:
```diff
@@ -147,6 +148,15 @@ message TlsSessionTicketKeys {
+  string provider_name = 1 [(validate.rules).string.min_bytes = 1];
+
+  // Private key operations provider specific configuration.
+  google.protobuf.Any typed_config = 2;
```
I think the answer to this thread is we should support both, since `Any` is not production ready yet and `Struct` has been undeprecated.

htuch commented on api/envoy/api/v2/auth/cert.proto:
```diff
@@ -315,6 +325,8 @@ message CommonTlsContext {
 
   reserved 5;
+
+  PrivateKeyOperations private_key_operations = 9;
```
This seems effectively a `oneof` with `TlsCertificate.private_key`. We can't promote it to a `oneof` without breaking the API, but should this field live next to that in this file and have a comment indicating the exclusive nature of these fields?

htuch replied:
We can't do `oneof` for wire compat reasons anyway, so you have flexibility here.

PiotrSikora replied:
I was under the impression that you can add `oneof` to non-repeatable fields and stay wire-compatible?

htuch replied:
Wire compatible yes, but it breaks Go language bindings. We've had a request from the gRPC team to avoid doing this gratuitously, this seems a reasonable accommodation to make. See https://github.com/envoyproxy/envoy/issues/6271.

PiotrSikora replied:
Ah, I wasn't aware of that, thanks! Yes, it makes sense to accommodate them (though, it's unfortunate that language bindings are source of this restriction).

ipuustin replied:
Ok, I'll move the private key operations provider to TlsCertificate. I'll have to set up the SSL socket bit differentlythen (because I'll need to initialize the private key methods for each `SSL_CTX`).

htuch commented on include/envoy/ssl/context_manager.h:
```diff
@@ -38,6 +39,12 @@ class ContextManager {
   virtual void iterateContexts(std::function<void(const Context&)> callback) PURE;
+
+  /**
+   * Access the private key operations manager, which is part of SSL
```
Again, I think I would prefer some naming that indicates this is specialized for BoringSSL only (or make this a top-level `union` of the two types).

htuch replied:
I agree that it is implicit and unambiguous, and other names would probably be longer and more cumbersome. Do you think we will maintain a BoringSSL only stance long term given the work that @bdecoste is doing?

htuch replied:
(maybe @bdecoste also wants to take a look at this discussion :D )

PiotrSikora replied:
Yes, the decision made at the EnvoyCon in December was that we're going to only support BoringSSL in-tree, and the approach that @bdecoste is taking (standalone extension vs compile-time alternative) is a direct result of that discussion.

cc @mattklein123 @lizan @bdecoste

htuch replied:
OK, then we can keep this naming convention, resolved, thanks.

htuch commented on include/envoy/ssl/private_key/private_key.h:
```diff
@@ -0,0 +1,82 @@
+namespace Envoy {
+namespace Server {
+namespace Configuration {
+class TransportSocketFactoryContext;
```
Is this to break an include dependency loop?

ipuustin replied:
Yes, that's why this is needed -- Bazel will complain if this is done with a real dependency.

htuch replied:
Arguably we could do some header refactoring to fix this, but I think this is OK if this is a one off.

ipuustin replied:
Ok. Just for the record, the dep loop would be this:

        //include/envoy/ssl:context_manager_interface
        //include/envoy/ssl:context_config_interface
    .-> //include/envoy/ssl:tls_certificate_config_interface
    |   //include/envoy/ssl/private_key:private_key_interface
    |   //include/envoy/server:transport_socket_config_interface
    |   //include/envoy/secret:secret_manager_interface
    |   //include/envoy/secret:secret_provider_interface
    `-- //include/envoy/ssl:tls_certificate_config_interface

htuch commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -124,6 +135,23 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
+  if (status == Envoy::Ssl::PrivateKeyOperationStatus::Success) {
+    ENVOY_CONN_LOG(debug, "async handshake complete", callbacks_->connection());
+    if (!handshake_complete_) {
+      // It's possible that the async call comes in later, but the handshare has been retried from
```
Nit: handshake

htuch commented:
@PiotrSikora do you want to  take a pass and do some design level feedback? I'm good with the high-level now, would be good to allow @ipuustin to commit on it for implementation.

PiotrSikora commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -124,6 +135,23 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
+
+  if (status == Envoy::Ssl::PrivateKeyOperationStatus::Success) {
+    ENVOY_CONN_LOG(debug, "async handshake complete", callbacks_->connection());
+    if (!handshake_complete_) {
```
Can this method be ever called with `handshake_complete_ == true`? Perhaps this should be an `ASSERT()` instead?

ipuustin replied:
Ok, changed to `ASSERT()`.

ipuustin replied:
I mean, changed `async_handshake_in_progress_` check to `ASSERT()`. I think it's possible that (depending on how the private key method provider is implemented) there's a race between the async callback and a `doHandshake()` call from `doWrite()` or `doRead()`. In that case it could be that the handshake is already completed and an ASSERT for `handshake_complete_` would be wrong.

PiotrSikora commented on source/extensions/transport_sockets/tls/ssl_socket.h:
```diff
@@ -89,6 +94,9 @@ class SslSocket : public Network::TransportSocket,
   mutable std::string cached_url_encoded_pem_encoded_peer_certificate_;
   mutable std::string cached_url_encoded_pem_encoded_peer_cert_chain_;
+  Envoy::Ssl::PrivateKeyOperationsProviderSharedPtr provider_;
+  Envoy::Ssl::PrivateKeyOperationsPtr ops_;
```
Do we need to store this?

ipuustin replied:
The provider is no longer needed. The `ops_` (or actually a vector now) is meant for connecting the operations lifecycle with the `SslSocket` lifecycle.

PiotrSikora commented on include/envoy/ssl/private_key/private_key_callbacks.h:
```diff
@@ -0,0 +1,30 @@
+   * @param status is "Success" or "Failure" depending on whether the private key operation was
+   * successful or not.
+   */
+  virtual void complete(PrivateKeyOperationStatus status) PURE;
```
I'm a bit worried about the case where the connection (and `SslSocket`) goes away while waiting for the async handshake to complete, and then the `complete()` callback is called, resulting in crash.

Do we have any protections against that?

ipuustin replied:
Originally the idea was that the `PrivateKeyOperations` lifecycle would be tied to the `SslSocket` lifecycle. I'll have to do this bit differently now when the operations are tied to certificates (which may change on-the-fly if multiple certificate support is used).

ipuustin commented:
I moved the private key method support into the `TlsCertificate` message (from `TlsContext`). Some things are not yet very well tested, such as if the multiple certificate support works as expected. I changed the naming conventions a bit, but I'll need to think about it still some more.

htuch commented on include/envoy/ssl/private_key/private_key.h:
```diff
@@ -0,0 +1,91 @@
+   * function parameter, so this enables the private key method provider
+   * to use SSL connection custom data fields instead.
+   */
+  virtual bool associateWithSsl(SSL* ssl) PURE;
```
Why isn't this part of `PrivateKeyOperationsCallbacks`? I.e. someone implementing the lifecycle callbacks around an async method should also be able to take an initial SSL binding call as well.

ipuustin replied:
I removed the `associateWithSsl()` function because the SSL object pointer was already available in the provider call. I also did a bunch of renaming effort which should hopefully move the naming conventions closer to the BoringSSL.

ipuustin commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -60,6 +60,17 @@ void SslSocket::setTransportSocketCallbacks(Network::TransportSocketCallbacks& c
+  provider_ = ctx_->getPrivateKeyOperationsProvider();
+  if (provider_) {
+    ops_ = provider_->getPrivateKeyOperations(*this, callbacks_->connection().dispatcher());
+    if (ops_) {
```
I changed the logic a bit. Originally the idea was that there might be a fallback to software processing, but now with the potential replacing of `private_key` with `private_key_method` this should be fatal.

ipuustin commented:
Thanks @htuch! My original idea was to get design-level comments and do the API design work here "in the open" so that other people who might want to use the API could participate in the design and tell if the API is suitable for their use cases.

When the design is more or less nailed down, I'll start writing proper tests (also an example SW-only private key method provider) and documentation. Currently with these items missing the PR isn't in a mergeable state yet. All review comments are fine at this point (nits and all), even though there might still be pretty large changes if the API has some drawbacks or could be implemented in a nicer way.

ipuustin commented:
I added a bunch of unit tests. I adapted code from BoringSSL reference RSA private key method tests, and I added the relevant BoringSSL license (BSD) to one file (test/extensions/transport_sockets/tls/rsa_private_key_method_provider.cc) accordingly. If this is not OK then I can write the provider from scratch (ECDSA maybe).

ipuustin commented:
@htuch Is it fine if I rebase this PR on top of the latest master branch (even if https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md asks not to)? There's a merge conflict and master has a lot of fixes which might help with the tests.

htuch commented:
@ipuustin you can in this case, but in general, the best practice is to do a fetch/master at your end and then push, that way you don't need to rewrite history. The basic heuristic is that before your first GH push and PR open, you can rebase freely, after that, you should be doing proper master branch merges and pushes (which don't rewrite history).

ipuustin commented:
@htuch Thanks, I did as you suggested and merged master into this branch. It helped to fix the testing.

htuch commented on test/extensions/transport_sockets/tls/rsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,296 @@
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+ * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE. */
```
https://github.com/envoyproxy/envoy/blob/master/source/common/chromium_url/LICENSE is how we've handled this for Chromium URL. It's somewhat nicer to have a standalone license file for tooling purposes. I'm not sure what the implication of this particular license or borrowing from BoringSSL is though; in general we have a BoringSSL dependency, so likely have a compatible license, but reading the BoringSSL license makes my head hurt - https://github.com/google/boringssl/blob/master/LICENSE. IANAL; @caniszczyk do you have thoughts on whether we can borrow code in this way?

htuch replied:
This is one I'd still like to close out.

caniszczyk replied:
the BoringSSL license(s) are fine for CNCF, just if you use any code, you have to keep all attributions in the header please

ipuustin replied:
I decided to rewrite the offending signing code using lower-level RSA methods. I have a tentative patch already, will test it a bit more before submitting.

htuch commented on test/extensions/transport_sockets/tls/rsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,296 @@
+
+  private_key_ = factory_context.api().fileSystem().fileReadToEnd(private_key_path);
+
+  method_ = std::make_shared<SSL_PRIVATE_KEY_METHOD>();
```
Is there any way to reuse the BoringSSL test code directly? I.e. via a `#include`?

ipuustin replied:
I'll investigate this. The problem is that we want to inject errors to the methods to test the various failure modes.

htuch replied:
It might be possible to make changes to BoringSSL and then bump our dependency version. The maintainers are Googlers, so we can try and coordinate on this if necessary.

ipuustin commented on api/envoy/api/v2/auth/cert.proto:
```diff
@@ -102,6 +104,19 @@ message TlsParameters {
+  oneof config_type {
+    google.protobuf.Struct config = 2;
+
+    google.protobuf.Any typed_config = 4;
```
It's a copy-paste error from the move to cert. I'll fix this.

ipuustin commented:
@htuch I think this PR is now in a shape where it can be reviewed. I don't plan of doing any more changes to the API unless I'm asked to. :-)

The PR still lacks the documentation but has now a bunch of tests, which already helped to iron out some issues. I think it doesn't make sense to write the docs before things are pretty much ready, because the docs will be more coherent if done when things are settled.

What do you have in mind when you say that you would like to do the review in chunks? Would you like me to split the PR or maybe refactor the commits so that they only depend on the previous ones?

htuch commented:
@ipuustin Ack. I had a quick skim again and I can't spot a clear way to break this up into sub-PRs, so let's review it as is.

@PiotrSikora would you be able to give this a full implementation review? I'll take a pass in the next couple of days (I usually do 24H SLO for reviews, but this warrants some extra scrutiny given the security nature and given the PRs existing cadence I think we can spread this out).

htuch commented:
(also CC @mergeconflict who has some interest in understanding SSL extension points in Envoy)

htuch commented on include/envoy/ssl/private_key/private_key.h:
```diff
@@ -0,0 +1,83 @@
+   * @param dispatcher supplies the owning thread's dispatcher.
+   * @return the private key operations instance.
+   */
+  virtual PrivateKeyConnectionPtr getPrivateKeyConnection(SSL* ssl,
```
Why is this called `PrivateKeyConnection`? I.e. what is the connection aspect? Connection is a heavily overloaded term in Envoy. Also, what are the object lifetime issues here and the purpose of this method? Who uses it? Etc. Having some clarify around this would help improve the readability.

ipuustin replied:
The idea was that this part corresponds to a single connection. I agree the naming isn't very good here. I'll try to come up with a better name.

ipuustin replied:
The functionality could be contained in the provider. I remember thinking about that approach too, but this seemed bit more straightforward because of the memory management (it's not possible that the callback comes after the socket was destroyed) and easier thinking about thread data access in the provider. If we remove the `PrivateKeyConnection` object we would have an API something like this:
```c++
// throws an error if the registration fails
void registerPrivateKeyConnection(SSL* ssl, PrivateKeyConnectionCallbacks& cb, Event::Dispatcher& dispatcher) PURE;

void unregisterPrivateKeyConnection(SSL* ssl) PURE;
```
Internally the provider would need to keep track of registered SSL objects (which it just borrows from the SSL sockets). A benefit of this approach would be avoiding of the naming issue. :-) I think I'll experiment a bit with this to get a better idea how it will work.

PiotrSikora replied:
> it's not possible that the callback comes after the socket was destroyed

I don't think that's true (but I agree that the current model would be acceptable if it was).

In a security-oriented use case, the private key will be loaded in a separate process, or even on a separate host, which means that you can receive result of the computation long after the socket was destroyed.

htuch commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -302,40 +302,56 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
+          private_key_method_provider->getBoringSslPrivateKeyMethod();
+      if (private_key_method == nullptr) {
+        throw EnvoyException(
+            fmt::format("Failed to get BoringSsl private key method from provider"));
```
Is there a test for this?

ipuustin replied:
I added a test for this now.

htuch commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -477,6 +493,19 @@ void ContextImpl::logHandshake(SSL* ssl) const {
+
+  for (uint i = 0; i < tls_contexts_.size(); i++) {
+    Envoy::Ssl::PrivateKeyMethodProviderSharedPtr provider =
+        tls_contexts_[i].getPrivateKeyMethodProvider();
```
How does multiple providers per context work?

ipuustin replied:
I got what you mean now -- I was first thinking you meant multiple certificates support. The code to tie the providers to the contexts is a remnant from the original design when the providers were context-specific. I need to figure out the best way to sort this out.

ipuustin replied:
It turns out that the code was correct after all. The `tls_contexts_` vector refers to `SSL_CTX` objects associated with a given context, and not to Envoy-level contexts as I first thought.

The function here gathers all private key method providers associated with a given TLS context into a vector which is then passed to `ssl_socket.cc`. The code in the ssl socket associates all the providers in the list with the ssl socket, so that if the `SSL_CTX` is changed during handshake in the multi-cert case, the callback to ssl socket will still work.

I added some tests (and a very limited ECDSA private key method provider) to validate this case. If you have any particular configuration in mind which should be tested, please tell, and I'll try it out and add a test for it.

htuch commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -124,6 +141,36 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
 }
 
+void SslSocket::complete() {
+  ASSERT(async_handshake_in_progress_);
```
I think it could be interesting to ASSERT that we're on the same thread that we were when we started the private key operation.

ipuustin replied:
I added now support for this. I had to add a function to `Dispatcher` public API, because it was bit tricky to get access to the thread id from `ssl_socket.cc`. What do you think: is this a too big change for the gain?

htuch commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -124,6 +141,36 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
+      // It may be semantically possible that the handshake is expecting read() or write(), so
+      // we shouldn't require that the handshake is complete. However, there can't be a second
+      // private key operation requested.
+      if (callbacks_->connection().state() == Network::Connection::State::Open) {
```
This code is a bit tricky, can you talk about how this is tested and whether there is a way to simplify? Would definitely want to hear from @PiotrSikora on this.

ipuustin replied:
Yes, this is quite complex. The idea here is that we can't leave handshakes hanging. Either they need to be completed or they need to error out. We only get the error result when `ssl_do_handshake()` is called the second time and the internal completion function in private key method is run. As you say, this function is quite crucial because the other end of the SSL connection can partially control the handshake and connection status. A bug in the logic could result in the connection never being closed or successful, leaving Envoy resources tied. @PiotrSikora, do you have any comments or opinions here?

If the second call to `ssl_do_handshake()` returns an error, we need to tell Envoy that the handshake isn't going anywhere, and this is why `callbacks_->connection().close()` is called. The problem with this is that this results in a `LocalClose` event, but synchronous failing results in `RemoteClose` event. It seems that this can't be easily fixed without adding a new parameter to `close()` call, which I didn't want to do to keep the PR scope more limited.

Testing the callback function is not very straightforward, because it depends on internal SSL connection state and the Envoy connection state machine. The current tests for success and failure cases are in `ssl_socket_test.cc`. All the cases which expect a `LocalClose` event are testing the error cases, for example this: https://github.com/envoyproxy/envoy/pull/6326/files#diff-be9d0f766b258e0940a6d9b1d35a7878R4325. Maybe the ssl_socket test harness could be extended to have better control over client connection state.

agl commented:
Just jumping in here to highlight that it's important to also handle session-ticket encryption keys if seeking to keep secrets out-of-process. For TLS 1.2, the ticket-encryption keys are probably _more_ sensitive than the certificate private key: given widespread ECDHE support, the private key doesn't allow global, passive decrypt, but the ticket-encryption keys _do_.

While, of course, not everything has to be handled in a single PR, I think it important to sign-post that simply moving private keys out-of-process may very well not achieve the intended goals.

Ticket-encryption keys can be abstracted using [this API](https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#ssl_ticket_aead_method_st) in BoringSSL.

agl commented on test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,180 @@
+  (void)out;
+  (void)out_len;
+  (void)signature_algorithm;
+  unsigned char hash[EVP_MAX_MD_SIZE];
```
(Seems really odd to declare all variables at the top, C89-style, but maybe Envoy does this?)

agl commented on test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,180 @@
+    return ssl_private_key_failure;
+  }
+
+  ec_key = ops->getPrivateKey();
```
there is `bssl::UniquePtr<EC_KEY>`, which will save you from having to remember to `EC_KEY_free` at each return point.

ipuustin replied:
Ok, that's a good tip.

agl commented on test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,180 @@
+    return ssl_private_key_failure;
+  }
+
+  ops->out_ = static_cast<uint8_t*>(OPENSSL_malloc(max_out));
```
You're using `OPENSSL_malloc` for ops&rarr;out_, but you don't need to because `privateKeyComplete` copies the value then frees it. So it could just be a std::vector or such to avoid the manual memory management. (I'm assuming that Envoy uses std::vector.)

ipuustin replied:
All right. I guess I was thinking the memory clearing would be good to have, but Vector looks nicer and leads to less code.

agl commented on test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,180 @@
+    return ssl_private_key_failure;
+  }
+
+  memcpy(out, ops->out_, ops->out_len_);
```
Should check that the amount of data copied is &le; `max_out`.

ipuustin replied:
Yes, that's right. Fixed.

agl commented on test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,180 @@
+
+  std::string private_key_path;
+
+  if (EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index == -1) {
```
(This assumes that no two threads call this function at the same time. Could well be true, just highlighting the assumption.)

ipuustin replied:
In my understanding this is true, because the providers are initialized only from the main thread as the TLS context configuration is processed. If there are several providers in the TLS context configuration (two certificates) the initialization is done sequentially.

agl commented on test/extensions/transport_sockets/tls/rsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,279 @@
+  // Add RSA padding to the the hash.
+  if (SSL_is_signature_algorithm_rsa_pss(signature_algorithm)) {
+    msg_len = RSA_size(rsa);
+    msg = static_cast<uint8_t*>(OPENSSL_malloc(msg_len));
```
(The comments from the ECDSA code largely apply here too.)

ipuustin replied:
Check.

davidben commented on test/extensions/transport_sockets/tls/rsa_private_key_method_provider.h:
```diff
@@ -0,0 +1,95 @@
+  RsaPrivateKeyConnection(SSL* ssl, Ssl::PrivateKeyConnectionCallbacks& cb,
+                          Event::Dispatcher& dispatcher, bssl::UniquePtr<EVP_PKEY> pkey,
+                          RsaPrivateKeyConnectionTestOptions& test_options);
+  RSA* getPrivateKey() { return EVP_PKEY_get0_RSA(pkey_.get()); }
```
I guess this is an example provider, so maybe it's intentional that you're breaking it down this far, but if you just use the `EVP_PKEY`, the API at that layer is a lot less work. You don't need to call `RSA_add_pkcs1_prefix` or assemble the PSS padding payload yourself or anything. (We usually do not recommend people assemble crypto operations from raw materials like that. It's more work and you sometimes miss out on hardening work that we can only do when we handle the whole thing.)

davidben replied:
Example:
https://cs.chromium.org/chromium/src/net/ssl/test_ssl_private_key.cc?rcl=a1c9e2c4a01f41cbb9af5b78feda358ca154bb1b&l=41

ipuustin replied:
Actually the first version I wrote copied the private key method sign code from the BoringSSL tests. However, it meant that the file was licensed using the BoringSSL BSD license, which would have made Envoy licensing more complex. If you scroll up this PR you can see discussion about the subject. I eventually decided to rewrite the sign function using lower-level operations to avoid the code copying.

BTW, have you considered adding external symbols for basic private key methods (RSA, ECDSA) to BoringSSL? It would seem that everyone writing their apps with private key method support needs to implement the basic SW-only private key methods for testing, before going on to write their own methods (TPM etc). Ideally these built-in BoringSSL private key methods would have some way to inject various errors for testing the application logic. The symbols could be behind some BoringSSL build flag (`--enable-private-key-method-testing`) or maybe in a separate test library.

davidben replied:
It seems the root cause here is not that BoringSSL isn't providing APIs, but that the PR has gone out of its way to not use the APIs we provide, at several layers. Even with the low-level ones, this should be using `RSA_sign` and `RSA_sign_pss_mgf1` so we'd take care of the padding internally.  But ultimately it really should be using EVP. Per that Chromium link, signing is just a few function calls. It's documented here, and is more-or-less the same as OpenSSL, except we don't have the pesky -1 and -2 returns to worry about.
https://commondatastorage.googleapis.com/chromium-boringssl-docs/evp.h.html

I'm not sure what adding basic private key methods would accomplish. Installing them on the `SSL_CTX` would be exactly the same as not configuring an `SSL_PRIVATE_KEY_METHOD` at all, so you're not actually testing your project's `SSL_PRIVATE_KEY_METHOD` code. Rather, what you want is to test a custom implementation of your own project's offloading interface with a software-backed implementation. For that, you do not want an `SSL_PRIVATE_KEY_METHOD` but just convenience functions to do the operations on an `EVP_PKEY`. I'm not opposed to that, but, as noted, the EVP API is already pretty close and the complexity in this PR is a bit self-inflicted.

Re bug injection, that should be done in test code, not scattered throughout production code where it risks real bugs. If you want an operation to fail, just install something that fails. If you want it to give the wrong answer, give the wrong answer.

ipuustin commented:
@agl, @davidben: Thanks for the review comments! I'll try to answer them separately. However, the proposal to add support for ticket sealing and opening in the extension (using for instance a TPM) is a design consideration so I started investigating it first.

It appears that the sealing functionality fits pretty well with the current design because the calls are synchronous. One issue is that we need to think how it will coexist with the current `session_ticket_keys` parameter, which exists on `TlsContext` level instead of `TlsCertificate` level.

@htuch, do you think we should extend this PR to cover the ticket encryption in extensions? It seems that this will fit pretty well in the current mode (if we assume that out-of-process ticket encryption will be done by a private key method provider) l so we could also do this work in a follow-up PR. A WIP patch would look like this:

```patch
From e819150d4e85bc93d8afeac9e5591a0f77433dbb Mon Sep 17 00:00:00 2001
From: Ismo Puustinen <ismo.puustinen@intel.com>
Date: Fri, 31 May 2019 15:42:32 +0300
Subject: [PATCH] tls: provide external encryption for session tickets.

Add support for external sealing and opening of session tickets. This is
used only for cases where a BoringSSL private key method provider is
present and provides support for getBoringSslAeadMethod() operation.

The `session_ticket_keys` parameter is not considered if a private key
method provider is used, because the ticket encryption is handled by the
extension.

Signed-off-by: Ismo Puustinen <ismo.puustinen@intel.com>
---
 include/envoy/ssl/private_key/private_key.h   |  8 +++
 .../transport_sockets/tls/context_impl.cc     |  8 ++-
 .../tls/ecdsa_private_key_method_provider.h   |  1 +
 .../tls/rsa_private_key_method_provider.cc    | 57 +++++++++++++++++++
 .../tls/rsa_private_key_method_provider.h     |  2 +
 test/mocks/ssl/mocks.h                        |  1 +
 6 files changed, 76 insertions(+), 1 deletion(-)

diff --git a/include/envoy/ssl/private_key/private_key.h b/include/envoy/ssl/private_key/private_key.h
index 4cb15f199..0c20e2795 100644
--- a/include/envoy/ssl/private_key/private_key.h
+++ b/include/envoy/ssl/private_key/private_key.h
@@ -21,6 +21,7 @@ class TransportSocketFactoryContext;
 namespace Ssl {
 
 typedef std::shared_ptr<SSL_PRIVATE_KEY_METHOD> BoringSslPrivateKeyMethodSharedPtr;
+typedef std::shared_ptr<SSL_TICKET_AEAD_METHOD> BoringSslAeadMethodSharedPtr;
 
 class PrivateKeyConnection {
 public:
@@ -51,6 +52,13 @@ public:
    * configuration.
    */
   virtual BoringSslPrivateKeyMethodSharedPtr getBoringSslPrivateKeyMethod() PURE;
+
+  /**
+   * Get the aead methods from the provider.
+   * @return the aead methods associated with this provider and
+   * configuration.
+   */
+  virtual BoringSslAeadMethodSharedPtr getBoringSslAeadMethod() PURE;
 };
 
 typedef std::shared_ptr<PrivateKeyMethodProvider> PrivateKeyMethodProviderSharedPtr;
diff --git a/source/extensions/transport_sockets/tls/context_impl.cc b/source/extensions/transport_sockets/tls/context_impl.cc
index ca98245d0..8f1d1b441 100644
--- a/source/extensions/transport_sockets/tls/context_impl.cc
+++ b/source/extensions/transport_sockets/tls/context_impl.cc
@@ -307,7 +307,7 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
     // We either have a private key or a BoringSSL private key method provider.
     if (private_key_method_provider) {
       ctx.private_key_method_provider_ = private_key_method_provider;
-      // The provider has a reference to the private key method for the context lifetime.
+      // The provider has a reference to the private key and aead methods for the context lifetime.
       Ssl::BoringSslPrivateKeyMethodSharedPtr private_key_method =
           private_key_method_provider->getBoringSslPrivateKeyMethod();
       if (private_key_method == nullptr) {
@@ -315,6 +315,12 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
             fmt::format("Failed to get BoringSsl private key method from provider"));
       }
       SSL_CTX_set_private_key_method(ctx.ssl_ctx_.get(), private_key_method.get());
+
+      Ssl::BoringSslAeadMethodSharedPtr aead_method =
+          private_key_method_provider->getBoringSslAeadMethod();
+      if (aead_method != nullptr) {
+        SSL_CTX_set_ticket_aead_method(ctx.ssl_ctx_.get(), aead_method.get());
+      }
     } else {
       // Load private key.
       bio.reset(BIO_new_mem_buf(const_cast<char*>(tls_certificate.privateKey().data()),
diff --git a/test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.h b/test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.h
index 3ead45ceb..d166142fc 100644
--- a/test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.h
+++ b/test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.h
@@ -51,6 +51,7 @@ public:
                                                        Ssl::PrivateKeyConnectionCallbacks& cb,
                                                        Event::Dispatcher& dispatcher) override;
   Ssl::BoringSslPrivateKeyMethodSharedPtr getBoringSslPrivateKeyMethod() override;
+  Ssl::BoringSslAeadMethodSharedPtr getBoringSslAeadMethod() override { return nullptr; };
 
   static int ssl_ecdsa_connection_index;
 
diff --git a/test/extensions/transport_sockets/tls/rsa_private_key_method_provider.cc b/test/extensions/transport_sockets/tls/rsa_private_key_method_provider.cc
index 7fb587650..4eaadfedb 100644
--- a/test/extensions/transport_sockets/tls/rsa_private_key_method_provider.cc
+++ b/test/extensions/transport_sockets/tls/rsa_private_key_method_provider.cc
@@ -206,6 +206,58 @@ RsaPrivateKeyMethodProvider::getBoringSslPrivateKeyMethod() {
   return method_;
 }
 
+static size_t aeadMaxOverhead(SSL *ssl) {
+  (void) ssl;
+  return 0;
+}
+
+static int aeadSeal(SSL *ssl, uint8_t *out, size_t *out_len, size_t max_out_len,
+              const uint8_t *in, size_t in_len) {
+  (void) ssl;
+  size_t i;
+
+  if (max_out_len < in_len) {
+    return 0;
+  }
+
+  memset(out, 0, max_out_len);
+
+  // Fake encryption by flipping every byte.
+  for (i = 0; i < in_len; i++) {
+    out[i] = ~in[i];
+  }
+
+  *out_len = in_len;
+
+  return 1;
+}
+
+static enum ssl_ticket_aead_result_t aeadOpen(SSL *ssl, uint8_t *out, size_t *out_len,
+                                        size_t max_out_len, const uint8_t *in,
+                                        size_t in_len) {
+  (void) ssl;
+  size_t i;
+
+  if (max_out_len < in_len) {
+    return ssl_ticket_aead_error;
+  }
+
+  memset(out, 0, max_out_len);
+
+  // Fake decryption by flipping every byte.
+  for (i = 0; i < in_len; i++) {
+    out[i] = ~in[i];
+  }
+
+  *out_len = in_len;
+
+  return ssl_ticket_aead_success;
+}
+
+Ssl::BoringSslAeadMethodSharedPtr RsaPrivateKeyMethodProvider::getBoringSslAeadMethod() {
+  return aead_;
+};
+
 RsaPrivateKeyConnection::RsaPrivateKeyConnection(SSL* ssl, Ssl::PrivateKeyConnectionCallbacks& cb,
                                                  Event::Dispatcher& dispatcher,
                                                  bssl::UniquePtr<EVP_PKEY> pkey,
@@ -272,6 +324,11 @@ RsaPrivateKeyMethodProvider::RsaPrivateKeyMethodProvider(
   method_->sign = privateKeySign;
   method_->decrypt = privateKeyDecrypt;
   method_->complete = privateKeyComplete;
+
+  aead_ = std::make_shared<SSL_TICKET_AEAD_METHOD>();
+  aead_->max_overhead = aeadMaxOverhead;
+  aead_->seal = aeadSeal;
+  aead_->open = aeadOpen;
 }
 
 } // namespace PrivateKeyMethodProvider
diff --git a/test/extensions/transport_sockets/tls/rsa_private_key_method_provider.h b/test/extensions/transport_sockets/tls/rsa_private_key_method_provider.h
index f6b3afd52..35432777c 100644
--- a/test/extensions/transport_sockets/tls/rsa_private_key_method_provider.h
+++ b/test/extensions/transport_sockets/tls/rsa_private_key_method_provider.h
@@ -67,11 +67,13 @@ public:
                                                        Ssl::PrivateKeyConnectionCallbacks& cb,
                                                        Event::Dispatcher& dispatcher) override;
   Ssl::BoringSslPrivateKeyMethodSharedPtr getBoringSslPrivateKeyMethod() override;
+  Ssl::BoringSslAeadMethodSharedPtr getBoringSslAeadMethod() override;
 
   static int ssl_rsa_connection_index;
 
 private:
   Ssl::BoringSslPrivateKeyMethodSharedPtr method_{};
+  Ssl::BoringSslAeadMethodSharedPtr aead_{};
   std::string private_key_;
   RsaPrivateKeyConnectionTestOptions test_options_;
 };
diff --git a/test/mocks/ssl/mocks.h b/test/mocks/ssl/mocks.h
index f2d3c624e..22f9601bd 100644
--- a/test/mocks/ssl/mocks.h
+++ b/test/mocks/ssl/mocks.h
@@ -82,6 +82,7 @@ public:
                PrivateKeyConnectionPtr(SSL* ssl, PrivateKeyConnectionCallbacks& cb,
                                        Event::Dispatcher& dispatcher));
   MOCK_METHOD0(getBoringSslPrivateKeyMethod, BoringSslPrivateKeyMethodSharedPtr());
+  MOCK_METHOD0(getBoringSslAeadMethod, BoringSslAeadMethodSharedPtr());
 };
 
 } // namespace Ssl
-- 
2.21.0
```

htuch commented:
@ipuustin sounds like a good plan to split into a future PR, I think the high order point is that we should support this for a complete solution. This PR is already quite big, so definitely happy to see more punted to followsup if possible.

PiotrSikora commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -60,6 +60,30 @@ void SslSocket::setTransportSocketCallbacks(Network::TransportSocketCallbacks& c
+    // We keep track of the private key operations for memory management purposes (the operations
+    // object if destroyed when the SslSocket is destroyed). The operations objects are unique
+    // because they need to associated with the SSL objects so that user data can be passed to the
+    // BoringSSL private key methods.
```
`SSL` object is passed to each of the `SSL_PRIVATE_KEY_METHOD` methods, so you can access `ex_data` without creating unique object for each connection.

ipuustin replied:
Yes. The wording in the comment could be better. There is no "need" of the association, but it is convenient. I'll postpone fixing this a bit if we decide to leave the connection out of the API.

PiotrSikora commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -124,6 +148,37 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
   return {action, bytes_read, end_stream};
 }
 
+void SslSocket::complete() {
```
Nit: It would be more readable if this method would be called `onPrivateMethodComplete` or something similar.

ipuustin replied:
That makes sense. Will fix.

PiotrSikora commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -238,7 +299,7 @@ Network::IoResult SslSocket::doWrite(Buffer::Instance& write_buffer, bool end_st
 
 void SslSocket::shutdownSsl() {
-  ASSERT(handshake_complete_);
+  ASSERT(handshake_complete_ || async_handshake_in_progress_);
```
If there is a async handshake in progress, then we should notify `PrivateKeyConnection` or `PrivateKeyMethodProvider` that this connection is going away and that we're no longer interested in the result, so that more complex providers can cancel the costly computation.

Probably in `SslSocket::closeSocket` and not here, though.

ipuustin replied:
Yes. The original idea was to tie the object lifecycles so that the `PrivateKeyConnection` destructor can cancel the operation. It might be better to do it explicitly though to prevent any races.

ipuustin replied:
@PiotrSikora: I added the unregistering now to `SslSocket::closeSocket()`. However, I'm not sure if the state machine will always call it; it seems that `SslSocket::doWrite()` can directly call `shutdownSsl()` too. Would it be safer that unregistering would be part of `SslSocket` destructor, and we would just ignore any async callbacks after `closeSocket()` and before the object destruction?

PiotrSikora replied:
I believe it will always call `closeSocket()`.

ipuustin replied:
Ok, this is good.

PiotrSikora commented on test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,180 @@
+    SSL* ssl, Ssl::PrivateKeyConnectionCallbacks& cb, Event::Dispatcher& dispatcher) {
+  bssl::UniquePtr<BIO> bio(
+      BIO_new_mem_buf(const_cast<char*>(private_key_.data()), private_key_.size()));
+  bssl::UniquePtr<EVP_PKEY> pkey(PEM_read_bio_PrivateKey(bio.get(), nullptr, nullptr, nullptr));
```
There is really no need to process and create new instance of the private key for each connection, one instance per provider should be enough.

ipuustin replied:
Good catch! Will fix.

ipuustin commented on api/envoy/api/v2/auth/cert.proto:
```diff
@@ -118,6 +133,10 @@ message TlsCertificate {
+
+  // BoringSSL private key operations. This is an alternative to private_key field. This can't be
+  // marked as "oneof" due to API compatibility reasons.
+  PrivateKeyMethod private_key_method = 6;
```
Ok.

ipuustin commented on source/common/ssl/tls_certificate_config_impl.cc:
```diff
@@ -22,9 +23,37 @@ TlsCertificateConfigImpl::TlsCertificateConfigImpl(
-  if (certificate_chain_.empty() || private_key_.empty()) {
+TlsCertificateConfigImpl::TlsCertificateConfigImpl(
+    const envoy::api::v2::auth::TlsCertificate& config, Api::Api& api,
+    bool expect_private_key_method)
```
The reason for the private constructor was an attempt to avoid copying the initialization code into the two public constructors. When the constructor is called from SDS initialization, the error handling logic goes bit differently, because there is no possibility to set the private key method (the `expect_private_key_method` value will be false) and the factory context isn't available. The separate constructor was needed because I wanted to have factory_context as a reference.

But you're right that the logic is complex. I'll look if I can make it simpler.

ipuustin commented on source/extensions/transport_sockets/tls/context_config_impl.cc:
```diff
@@ -26,7 +26,9 @@ std::vector<Secret::TlsCertificateConfigProviderSharedPtr> getTlsCertificateConf
     std::vector<Secret::TlsCertificateConfigProviderSharedPtr> providers;
     for (const auto& tls_certificate : config.tls_certificates()) {
-      if (!tls_certificate.has_certificate_chain() && !tls_certificate.has_private_key()) {
+      Ssl::PrivateKeyMethodProviderSharedPtr private_key_method_provider = nullptr;
```
Right. Fixed.

ipuustin commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -302,40 +302,56 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
+          private_key_method_provider->getBoringSslPrivateKeyMethod();
+      if (private_key_method == nullptr) {
+        throw EnvoyException(
+            fmt::format("Failed to get BoringSsl private key method from provider"));
```
Ok.

ipuustin commented on source/extensions/transport_sockets/tls/context_impl.cc:
```diff
@@ -302,40 +302,56 @@ ContextImpl::ContextImpl(Stats::Scope& scope, const Envoy::Ssl::ContextConfig& c
+        throw EnvoyException(
+            fmt::format("Failed to get BoringSsl private key method from provider"));
+      }
+      SSL_CTX_set_private_key_method(ctx.ssl_ctx_.get(), private_key_method.get());
```
I can add that, sure. Will it just be a boolean return value or can there be a more elaborate situations with the keys which we would need to handle?

PiotrSikora replied:
`bool` is fine (private key is either FIPS-worthy or not).

ipuustin commented:
/retest

Added a bunch of fixes based on the review feedback. The bigger refactorings (removing `PrivateKeyMethodConnection`, fixing the three-constructor problem in `TlsCertificateConfigImpl`, going back to the high-level crypto APIs) are still WIP.

ipuustin commented:
@PiotrSikora: I have now commits which should address the latest round of review feedback. `PrivateKeyMethodConnection` class was removed and the complex constructor setup was made simpler. The `checkFips()` function was added to the API and used when setting up the TLS context.

ipuustin commented:
/retest

ipuustin commented:
Let's try the Mac test again...

/retest

rojkov commented:
/azp run

lizan commented:
@ipuustin can you merge master?

lizan commented on include/envoy/ssl/private_key/private_key.h:
```diff
@@ -0,0 +1,91 @@
+
+class PrivateKeyOperations {
+public:
+  virtual ~PrivateKeyOperations() {}
```
`= default` for here and all destructors in throughout this PR.

ipuustin replied:
Ok.

lizan commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -165,12 +207,18 @@ PostIoAction SslSocket::doHandshake() {
+      async_handshake_in_progress_ = true;
       return PostIoAction::KeepOpen;
     default:
+      ENVOY_CONN_LOG(debug, "handshake error: {}", callbacks_->connection(), err);
```
:+1:

lizan commented on test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,188 @@
+static ssl_private_key_result_t privateKeySign(SSL* ssl, uint8_t* out, size_t* out_len,
+                                               size_t max_out, uint16_t signature_algorithm,
+                                               const uint8_t* in, size_t in_len) {
+  (void)out_len;
```
nit: use UNREFERENCED_PARAMETER macro or just make those params unnamed.

ipuustin replied:
Ok, made them unnamed.

lizan commented on test/extensions/transport_sockets/tls/rsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,258 @@
+static ssl_private_key_result_t privateKeySign(SSL* ssl, uint8_t* out, size_t* out_len,
+                                               size_t max_out, uint16_t signature_algorithm,
+                                               const uint8_t* in, size_t in_len) {
+  RSA* rsa;
```
nit: declare RSA and EVP_MD when it is needed and inline below

ipuustin replied:
Fixed.

ipuustin commented:
@lizan: merged master again now.

ipuustin commented:
/retest

ipuustin commented:
Following the review comments is bit tricky, but I think the proposed fixes have all been made. I went through the comments and it seems that the current opens are more or less these:

1. @PiotrSikora: do you have any comments regarding the handshake logic here: https://github.com/envoyproxy/envoy/pull/6326/files#diff-48da3a44e881b026eafba5f129e46ed2R169
2. @PiotrSikora: do you have any input regarding the connection state machine: should we release private key method operation resources in `closeSocket()` or `shutdownSSL()`? Is `closeSocket()` guaranteed to be called by Envoy for every `SslSocket` instance?
3. @htuch: do you think adding the `getCurrentThreadId()` function to https://github.com/envoyproxy/envoy/pull/6326/files#diff-e855885dc4595f9b0dda60f65e8ef8bfR206 is worth it? It's as far as I understand needed if we want to check that the callback is called in the correct thread context.
4. @htuch: do you think the current docs are enough? We can document the individual extensions as they come (such as the HSM extension).

If there are any other points which I missed please add. :-)

PiotrSikora commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -150,6 +161,37 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
+        // private key method handshake failed. If the connection state would be something else,
+        // the socket was already closed while waiting for async handshake.
+        ENVOY_CONN_LOG(debug, "async handshake completion error", callbacks_->connection());
+        drainErrorQueue();
```
`drainErrorQueue()` is already called in the only codepath that returns `PostIoAction::Close`.

PiotrSikora commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -150,6 +161,37 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
+  async_handshake_in_progress_ = false;
+
+  ENVOY_CONN_LOG(debug, "async handshake complete", callbacks_->connection());
+  if (!handshake_complete_) {
```
I don't think we could ever reach this codepath with `handshake_complete_ == true`.

ipuustin commented on source/common/ssl/tls_certificate_config_impl.cc:
```diff
@@ -30,5 +31,34 @@ TlsCertificateConfigImpl::TlsCertificateConfigImpl(
   }
```
It's true that if we convert the `TransportSocketFactoryContext` reference to a pointer and back we can reuse the same constructor. I guess whether or not that's better than having multiple constructors is a matter of taste and I'm fine either way. :-) I'll do the fix.

ipuustin commented on source/extensions/transport_sockets/tls/context_manager_impl.h:
```diff
@@ -33,11 +36,15 @@ class ContextManagerImpl final : public Envoy::Ssl::ContextManager {
   void removeEmptyContexts();
   TimeSource& time_source_;
   std::list<std::weak_ptr<Envoy::Ssl::Context>> contexts_;
+  PrivateKeyMethodManagerImpl private_key_operations_manager_{};
```
Ok.

ipuustin commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -150,6 +161,37 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
+    // It's possible that the async call comes in later, but the handshake has been retried from
+    // doWrite or similar. In that case the error handling has been done synchronously. */
+    PostIoAction result = doHandshake();
+    if (async_handshake_in_progress_ || result != PostIoAction::KeepOpen) {
```
The code was trying to prepare for two possible issues in the providers.

The first case was someone returning `ssl_private_key_retry` from the completion function even **after** being called from `onPrivateKeyMethodComplete()` callback. In this case the `async_handshake_in_progress_` flag would be become `true` again. The idea was that we could detect the error even though the resulting action would be `PostIoAction::KeepOpen`. However, as you say, we can just agree that providers are not allowed to return `ssl_private_key_retry` if they have already called the callback. Anyway in my opinion we shouldn't make the callback function idempotent if the operation is still pending, but request that it must be called exactly once.

The second case was to guard about the race condition in the provider where the key operation is ready but `doHandshake()` is called from `doWrite()` instead of `onPrivateKeyMethodComplete()` callback. This is why the `if (handshake_complete_)` path was there. However, since we are directly calling the handshake function again from the callback, we can just say that it's an error if the provider returns `ssl_private_key_success` or `ssl_private_key_failure` from the completion function **before** it has called the callback.

If we go with these (very reasonable) semantics then we can very well have the simpler form of the callback which you suggest. The example providers already work this way. I'll do the fix.

ipuustin commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -150,6 +161,37 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
+      // It may be semantically possible that the handshake is expecting read() or write(), so
+      // we shouldn't require that the handshake is complete. However, there can't be a second
+      // private key operation requested.
+      if (callbacks_->connection().state() == Network::Connection::State::Open) {
```
Ah, that's true. It's a benefit of the unregistration model.

ipuustin commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -150,6 +161,37 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
+        // Network::ConnectionEvent::LocalClose event.
+        // TODO(ipuustin): should there be a special event type for asynchronous close?
+        // Such as Network::ConnectionEvent::ConnectionFailed?
+        callbacks_->connection().close(Network::ConnectionCloseType::NoFlush);
```
Hmm, yes, I think you're right.

ipuustin commented:
/retest

PiotrSikora commented on test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,176 @@
+#include "test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.h"
```
I know that those are test providers, but we really don't need to have separate providers for ECDSA & RSA, this should be the same provider that can handle both types, since the differences are minimal.

This will also uncover issues with using static index to register private key connection on the `SSL` object.

ipuustin replied:
Ok, I'll merge the providers.

PiotrSikora commented on test/extensions/transport_sockets/tls/ecdsa_private_key_method_provider.cc:
```diff
@@ -0,0 +1,176 @@
+
+  if (EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index == -1) {
+    EcdsaPrivateKeyMethodProvider::ssl_ecdsa_connection_index =
+        SSL_get_ex_new_index(0, nullptr, nullptr, nullptr, nullptr);
```
You could use this instead:
```
int EcdsaPrivateKeyMethodProvider::sslConnectionIndex() {
  CONSTRUCT_ON_FIRST_USE(int, []() -> int {
    int ssl_connection_index = SSL_CTX_get_ex_new_index(0, nullptr, nullptr, nullptr, nullptr);
    RELEASE_ASSERT(ssl_connection_index >= 0, "");
    return ssl_connection_index;
  }());
}
```

ipuustin replied:
Ok, this looks nice.

PiotrSikora commented on source/extensions/transport_sockets/tls/ssl_socket.cc:
```diff
@@ -60,6 +60,17 @@ void SslSocket::setTransportSocketCallbacks(Network::TransportSocketCallbacks& c
   ASSERT(!callbacks_);
   callbacks_ = &callbacks;
 
+  // Associate this SSL connection with all the certificates (with their potentially different
```
The private key method is going to be called only on one of the configured providers, so registering it on all is a bit wasteful. Could we auto-register on `Sign()` / `Decrypt()` operations instead, somehow?

ipuustin replied:
We could maybe add a pointer to the `SslSocket` into the SSL object using `SSL_set_ex_data()` during SSL object creation, and then add a public function to `SslSocket` to get access to the thread `Dispatcher` object. Then that `SslSocket` pointer could be retrieved in `sign()`/`decrypt()` and `complete()` functions to get access to the dispatcher and to the callback object. Since there would be no pre-registration of provider to SSL object, we would probably need to add the private key method provider pointer to `SSL_CTX` object using `SSL_CTX_set_ex_data()`, so that provider configuration could be accessed from `sign()`/`decrypt()`.

However, there are some issues with this approach. Because we can't use references or smart pointers in SSL object user data functions, we will be reduced to using plain void pointers and casting. Following the program logic might become bit somewhat more complex too. Also, the maximum amount of certificates supported for a TLS context in the multi-cert use case is two, so we are dealing with small constants instead of an arbitrary number of certificates. Still, I can do the change if you believe this would be a good idea. What do you think?

PiotrSikora replied:
Let's leave it as-is. Thanks.

ipuustin commented on api/envoy/api/v2/auth/cert.proto:
```diff
@@ -102,13 +104,37 @@ message TlsParameters {
+  // <envoy_api_field_auth.TlsCertificate.private_key>` and :ref:`private_key_method
+  // <envoy_api_field_auth.TlsCertificate.private_key_method>` fields will result in an
+  // error.
+  PrivateKeyMethod private_key_method = 6;
```
I think `private_key_provider` could be good. It highlights that some external entity is handling the key logic. I'll change the name.

ipuustin commented:
@PiotrSikora the requested fixes should be done now. Can you take another look, please?

ipuustin commented:
/retest

ipuustin commented:
/azp run

ipuustin commented:
/retest Hmm, how can I rerun the macos tests...

ipuustin commented:
Ping for reviews to keep the stale bot away...

ipuustin commented:
@PiotrSikora @htuch I merged master again to solve the merge conflicts. Could you do another review pass, please?

ipuustin commented:
/retest Merged the master branch again to resolve conflicts. Changed `SslSocket` to use new dispatcher `isThreadSafe()` function instead of a custom thread safety check.

PiotrSikora commented:
I chatted with @ipuustin offline about this - I've been swamped with other stuff, but I'm going to do another pass later this week. Sorry for delay!

rojkov commented on source/extensions/transport_sockets/tls/ssl_socket.h:
```diff
@@ -101,6 +109,7 @@ class SslSocket : public Network::TransportSocket,
   mutable std::string cached_sha_256_peer_certificate_digest_;
   mutable std::string cached_url_encoded_pem_encoded_peer_certificate_;
   mutable std::string cached_url_encoded_pem_encoded_peer_cert_chain_;
+  bool async_handshake_in_progress_{};
```
Now `SslSocket` has 3 boolean flags `async_handshake_in_progress_`, `handshake_complete_`, `shutdown_sent_`.

@PiotrSikora @lizan would it make sense to have `SslSocketState` enum instead?

lizan replied:
Yeah that seems cleaner if works.

ipuustin commented:
@rojkov @lizan : I replaced the bools with an enum, please take a look. I also merged the master to get the tests working again.

@PiotrSikora : ping!

rojkov commented:
Thanks @ipuustin ! I'll update openssl tls accordingly.

htuch commented:
@lizan yeah, I think we can ship as soon as we have sign-off from SSL owners on this, which is @PiotrSikora (and optionally @agl / @davidben).

PiotrSikora commented on include/envoy/ssl/private_key/private_key.h:
```diff
@@ -0,0 +1,87 @@
+  virtual PrivateKeyMethodProviderSharedPtr
+  createPrivateKeyMethodProvider(const envoy::api::v2::auth::PrivateKeyProvider& message,
+                                 Envoy::Server::Configuration::TransportSocketFactoryContext&
+                                     private_key_method_provider_context) PURE;
```
Nit: `private_key_method_provider_context` -> `transport_socket_factory_context` or simply `factory_context`? (here and in other places)

PiotrSikora commented on include/envoy/ssl/private_key/private_key.h:
```diff
@@ -0,0 +1,87 @@
+   * Finds and returns a private key operations provider for BoringSSL.
+   *
+   * @param message a protobuf message object containing a
+   * PrivateKeyProvider message.
```
Nit: you can pull this into previous line.

PiotrSikora commented on include/envoy/ssl/private_key/private_key.h:
```diff
@@ -0,0 +1,87 @@
+   * @param message a protobuf message object containing a
+   * PrivateKeyProvider message.
+   * @param private_key_method_provider_context context that provides components for creating and
+   * initializing connections for keyless TLS etc.
```
Nit: `...connections using asynchronous private key operations.`

lizan commented:
Needs @htuch's approval (for requested changes and api)

ipuustin commented:
A big thanks to all reviewers!
